tests/functional/test_fast_deps.py

```diff
@@ -101,3 +103,31 @@ def test_hash_mismatch(script: PipTestEnvironment, tmp_path: pathlib.Path) -> No
         expect_error=True,
     )
     assert "DO NOT MATCH THE HASHES" in result.stderr
+
+
+@mark.network
+def test_hash_mismatch_existing_download(
+    script: PipTestEnvironment, tmp_path: pathlib.Path
+) -> None:
+    reqs = tmp_path / "requirements.txt"
+    reqs.write_text("idna==2.10")
+    dl_dir = tmp_path / "downloads"
+    dl_dir.mkdir()
+    idna_wheel = dl_dir / "idna-2.10-py2.py3-none-any.whl"
+    idna_wheel.write_text("asdf")
+    result = script.pip(
+        "download",
+        "--use-feature=fast-deps",
+        "-r",
+        str(reqs),
+        "-d",
+        str(dl_dir),
+        allow_stderr_warning=True,
+    )
+    assert re.search(
+        r"WARNING: Previously-downloaded file.*has bad hash", result.stderr
+    )
```

Comment:
I think we have something built into `script` that can do this directly

Reply:
Can’t find anything, maybe I’m delusional 💭

Reply:
I couldn't find any uses of `hash_file()` except in `tests/unit/test_wheel.py`, and the only uses of `hashlib` I found were to generate records for mock wheels. It probably wouldn't be a bad idea to build in some functionality to explicitly scan for a desired warning or error from pip output, though, instead of having to check it after with `allow_stderr_warning=True`.